src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -403,6 +404,23 @@ func Register(
 		sessionID = util.RandomString(sessionIDLength)
 	}
 
+	// Don't allow numeric usernames less than MAX_INT64.
+	if _, err := strconv.ParseInt(r.Username, 10, 64); err == nil {
+		return util.JSONResponse{
+			Code: http.StatusBadRequest,
+			JSON: jsonerror.InvalidUsername("Numeric user IDs are reserved"),
+		}
+	}
+	// Auto generate a numeric username if r.Username is empty
+	if r.Username == "" {
+		id, err := accountDB.GetNewNumericLocalpart(req.Context())
+		if err != nil {
+			return jsonerror.InternalServerError()
```

Comment:
We use `return httputil.LogThenError(req, err)` so that it includes an appropriate error message in the response.